src/event.rs

```diff
@@ -1,139 +1,270 @@
-// Copyright (c) 2018-2020  Brendan Molloy <brendan@bbqsrc.net>
+// Copyright (c) 2018-2021  Brendan Molloy <brendan@bbqsrc.net>,
+//                          Ilya Solovyiov <ilya.solovyiov@gmail.com>,
+//                          Kai Ren <tyranron@gmail.com>
 //
 // Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
 // http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
 // <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
-//! Key occurrences in the lifecycle of a Cucumber execution.
+
+//! Key occurrences in a lifecycle of a [Cucumber] execution.
 //!
-//! The top-level enum here is `CucumberEvent`.
+//! The top-level enum here is [`Cucumber`].
 //!
-//! Each event enum contains variants indicating
-//! what stage of execution Cucumber is at and,
-//! variants with detailed content about the precise
-//! sub-event
-
-pub use super::ExampleValues;
-use std::{fmt::Display, rc::Rc};
-
-/// The stringified content of stdout and stderr
-/// captured during Step execution.
-#[derive(Debug, Clone, PartialEq, Eq)]
-pub struct CapturedOutput {
-    pub out: String,
-    pub err: String,
-}
+//! Each event enum contains variants indicating what stage of execution
+//! [`Runner`] is at, and variants with detailed content about the precise
+//! sub-event.
+//!
+//! [`Runner`]: crate::Runner
+//!
+//! [Cucumber]: https://cucumber.io
+
+use std::{any::Any, sync::Arc};
+
+/// Alias for a [`catch_unwind()`] error.
+///
+/// [`catch_unwind()`]: std::panic::catch_unwind()
+pub type Info = Box<dyn Any + Send + 'static>;
+
+/// Top-level [Cucumber] run event.
+///
+/// [Cucumber]: https://cucumber.io
+#[derive(Debug)]
+pub enum Cucumber<World> {
+    /// Event for a [`Cucumber`] execution started.
+    Started,
 
-/// Panic source location information
-#[derive(Debug, Clone, PartialEq, Eq)]
-pub struct Location {
-    pub file: String,
-    pub line: u32,
-    pub column: u32,
+    /// [`Feature`] event.
+    Feature(Arc<gherkin::Feature>, Feature<World>),
+
+    /// Failed to parse a [`Feature`] file.
+    ParsingError(gherkin::ParseFileError),
```

Comment:
Not nice really. Better decoupled into `Result<Cucumber, ParsingError>`. It's better to lay a fallible path through the whole pipeline for better extensibility.